### PR TITLE
kubectl --wait flag

### PR DESCRIPTION
**Pull request by nilebox**
**What this PR does / why we need it**:
Adds an optional `--wait` flag to `kubectl delete` command.
When the flag is set to true and kubectl is doing cascading deletion, then it waits to ensure that the resource has been deleted before returning.
This PR is based off #46471 (not a complete copy, just idea and some functions) with a number of key differences:
- The `--wait` flag is explicit and is `false` by default
  - While it contradicts #42594, it makes the behavior more explicit and straightforward (no `nil` handling in the code, no exceptions for default behavior for namespace and pods)
  - When we rebase #59851 on this change, it will be up to the user whether to wait for all finalizers and GC to complete or not
  - After we remove client-side reapers and rely on GC foreground deletion instead, we will also be able to get rid of internal `WaitForGracefulDeletion` flag => there will be no more client-side deletion hacks
- The `--wait` flag doesn't apply to resources with reapers
  - In the #46471, the wait flag was always on for resources with reaper, which IMO is not consistent for when the user explicitly passes `--wait=false`, especially when there is a custom finalizer set on the resource?
  - The proper fix would probably be to pass the `waitForDeletion bool` parameter to the reaper's `Stop()` function, but I would like to discuss this first
  - Hopefully, reapers will soon go away and be replaced with foreground deletion
- Removed workarounds for backward compatibility with k8s 1.6
- Updated to match master

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
See #42594 and #59851

**Special notes for your reviewer**:

**Release note**:

```release-note

```

### CODE REVIEW

File: pkg/kubectl/delete.go [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright 2014 The Kubernetes Authors.
+Copyright 2018 The Kubernetes Authors.
```
Comment (nilebox) [inline pkg/kubectl/delete.go]: This GitHub diff is misleading. What I actually did:
- Renamed `delete.go` -> `reapers.go` to better reflect the code living in that file (hopefully soon to be deleted)
- Created a new `delete.go` file with utility functions for deletion.

**Comment by nilebox** [comment]
/sig CLI
@kubernetes/sig-cli-api-reviews

**Comment by nilebox** [comment]
I have made a commit in a separate branch adding support for `--wait` flag to reapers, see https://github.com/nilebox/kubernetes/commit/691012d080560a771a9cb5ee118ae94bec16f94f
It repeats reaper-related changes from #46471
@liggitt Please take a look and let me know if you think that this change is needed.

**Comment by soltysh** [comment]
/hold
I'd rather we clean the current implementation first, before introducing new functionality. That means I want https://github.com/kubernetes/kubernetes/pull/59851 in and reapers being removed first and then we can focus on this one.

**Comment by nilebox** [comment]
@soltysh that would be my preferred approach as well 👍

**Comment by nilebox** [comment]
@soltysh could you remove the `hold` label? See https://github.com/kubernetes/kubernetes/pull/59851#issuecomment-389960356
Otherwise those two PRs block each other.

**Comment by soltysh** [comment]
No, I want to keep the order:  #59851 gets in first, then https://github.com/kubernetes/kubernetes/pull/63979 only then this PR. I don't want to mess with the current code introduce weird logic which is removed in https://github.com/kubernetes/kubernetes/pull/63979. I want a clean and nice solution.

**Comment by soltysh** [comment]
Acutally, there's an alternative approach to this in https://github.com/kubernetes/kubernetes/pull/64034 that I prefer more, since it adds a generic wait mechanism that will be used in delete, but additionally provide a new command, which was asked back in the very early days of the project.

**Pull request closed without merging**